File: backend/app/api/permissions.py
# backend/app/api/permissions.py
"""
Role capabilities for the CV backend — LAYER 2.

This is the Python mirror of `frontend/app/lib/permissions.js`. The two files
MUST agree, and the shared capability names are what make a disagreement
visible: `zones.edit` means the same thing on both sides of the wire.

WHY THE BACKEND NEEDS THIS AT ALL

Zone writes do not go through a Next.js Server Action. The ZoneEditor talks
directly to FastAPI, so gating only the Next actions would leave
`POST /api/v1/zones/` reachable by any authenticated member regardless of role.
Before this file, a VIEWER's zone save reached the database and was stopped
only by RLS — which works, but returns "new row violates row-level security
policy for table zones" to a person who simply is not allowed to draw zones.

THE THREE LAYERS, AND WHICH ONE ACTUALLY HOLDS

    1. UI          hides the editor from a VIEWER    (MembersScreen / dashboard)
    2. this file   refuses the request with a 403    (endpoint guard)
    3. Postgres    zone_insert USING manage_org_ids()  ← the real boundary

Layer 3 is the one that cannot be bypassed, and it was measured rather than
assumed: a VIEWER inserting a zone is rejected with a row-level security
violation, while MANAGER and ADMIN succeed. Layer 2 exists to turn that into an
honest status code and a readable sentence, and to fail before touching the
database at all.

If this file and the RLS policies ever disagree, the policy is right and this
file is the bug.
"""

from typing import Optional

# capability -> roles that hold it. Kept in the same order and with the same
# names as the JS table, so the two can be diffed by eye.
CAPABILITIES = {
    # Reading — every ACTIVE member, enforced by user_org_ids().
    "dashboard.view": ("ADMIN", "MANAGER", "VIEWER"),
    "analytics.view": ("ADMIN", "MANAGER", "VIEWER"),
    "reports.export": ("ADMIN", "MANAGER", "VIEWER"),
    "zones.view": ("ADMIN", "MANAGER", "VIEWER"),
    "members.view": ("ADMIN", "MANAGER", "VIEWER"),
    # enforcedBy: employee_select — membership-based, like every other roster.
    "employees.view": ("ADMIN", "MANAGER", "VIEWER"),

    # Configuring the space — enforced by manage_org_ids().
    "zones.edit": ("ADMIN", "MANAGER"),
    "cameras.edit": ("ADMIN", "MANAGER"),
    "sites.edit": ("ADMIN", "MANAGER"),
    # Running analysis WRITES telemetry, so it is a configuration act rather
    # than a read. A VIEWER watching a live feed would be creating data.
    "analysis.run": ("ADMIN", "MANAGER"),
    # enforcedBy: employee_insert/update and soft_delete_employee(). Also gates
    # face enrolment, which is the act that makes a named person recognisable
    # by camera — the most consequential write in the product.
    "employees.edit": ("ADMIN", "MANAGER"),

    # Governing the organisation — enforced by admin_org_ids().
    "members.invite": ("ADMIN",),
    "members.manage": ("ADMIN",),
    "org.settings": ("ADMIN",),
}


def can(role: Optional[str], capability: str) -> bool:
    """
    May this role perform this capability?

    Fails CLOSED on an unknown role (None, or a suspended member with no active
    membership) AND on an unknown capability. The second matters: if a typo
    returned True, a misspelled capability would silently grant everyone
    everything and still look like working code.
    """
    allowed = CAPABILITIES.get(capability)
    if not allowed:
        return False
    if not isinstance(role, str):
        return False
    return role in allowed


def denial_message(capability: str) -> str:
    """
    The refusal, phrased so the reader knows who to ask.

    "You do not have permission" leaves someone stuck; "only an administrator or
    manager can do that" tells them what to do next.
    """
    allowed = CAPABILITIES.get(capability)
    if not allowed:
        return "That action is not available."
    if allowed == ("ADMIN",):
        return "Only an administrator can do that."
    if "MANAGER" in allowed and "VIEWER" not in allowed:
        return "Only an administrator or manager can do that."
    return "You do not have permission to do that."

File: frontend/app/lib/permissions.js
// frontend/app/lib/permissions.js
//
// THE SINGLE SOURCE OF TRUTH FOR WHAT EACH ROLE MAY DO.
//
// Three roles exist in the schema and are meaningless until something enforces
// them. This file is the one place that answers "may this role do X", so the
// answer cannot drift between a hidden button and the action behind it.
//
// WHERE THIS IS APPLIED — ALL THREE LAYERS, NOT JUST THE FIRST
//
//   1. UI            hide controls a role cannot use          (this file)
//   2. Server        re-check inside every action/endpoint    (this file)
//   3. Postgres RLS  the actual boundary                      (003_rls_policies)
//
// Layer 1 is courtesy: it stops a VIEWER staring at a button that will fail.
// Layer 2 is the application's own guard, because a Server Action is a POST
// endpoint the browser can call directly and it does not inherit the page's
// render decisions — hiding a button removes it from the screen, not from the
// network.
//
// Layer 3 is the one that actually holds. Verified by direct measurement, not
// assumed: a VIEWER inserting a zone gets
// "violates row-level security policy for table zones", while MANAGER and ADMIN
// succeed. If layers 1 and 2 were both deleted tomorrow, the data would still
// be safe; if only layer 3 were deleted, nothing else would save it.
//
// So this file must MIRROR the policies, never invent its own rules. Each
// capability below names the policy that really enforces it. When they
// disagree, the policy is right and this file is a bug.

/** Roles, most privileged first. Matches the `Role` enum in schema.prisma. */
export const ROLES = ['ADMIN', 'MANAGER', 'VIEWER'];

export const ROLE_LABELS = {
  ADMIN: 'Administrator',
  MANAGER: 'Manager',
  VIEWER: 'Viewer',
};

/**
 * capability -> roles that hold it.
 *
 * Read this table against 003_rls_policies.sql. The `enforcedBy` note on each
 * group is the policy or function that makes the rule real; this map only
 * decides what the interface offers.
 */
const CAPABILITIES = {
  // ── Reading. Every ACTIVE member of an org may read its measurements. ──
  // enforcedBy: user_org_ids() — membership-based, so it covers all three
  // roles and excludes platform operators entirely.
  'dashboard.view':   ['ADMIN', 'MANAGER', 'VIEWER'],
  'analytics.view':   ['ADMIN', 'MANAGER', 'VIEWER'],
  'reports.export':   ['ADMIN', 'MANAGER', 'VIEWER'],
  'zones.view':       ['ADMIN', 'MANAGER', 'VIEWER'],
  'members.view':     ['ADMIN', 'MANAGER', 'VIEWER'],
  // enforcedBy: employee_select — "orgId IN user_org_ids()", the same
  // membership-based read every other roster gets. The employee list is a
  // colleague list; what a VIEWER does not get is the editing controls.
  'employees.view':   ['ADMIN', 'MANAGER', 'VIEWER'],
  // Step 16's team comparison — everyone's measured figures in one table.
  //
  // enforcedBy: employee_day_stat_select (migration 022), which returns other
  // people's rows only to `manage_org_ids()` — ADMIN and MANAGER. A VIEWER
  // sees exactly one employee's figures: their own.
  //
  // So this capability is not the security boundary, the policy is; it exists
  // so a VIEWER is not shown a link to a table that would render with one row
  // and look broken. Hiding the link is a courtesy, not a control.
  'team.compare':     ['ADMIN', 'MANAGER'],

  // ── Configuring the space. ──
  // enforcedBy: manage_org_ids() — zone_insert/update/delete,
  // camera_insert/update/delete, site_insert/update.
  // A VIEWER is read-only by definition; a MANAGER runs the space.
  'zones.edit':       ['ADMIN', 'MANAGER'],
  'cameras.edit':     ['ADMIN', 'MANAGER'],
  'sites.edit':       ['ADMIN', 'MANAGER'],
  // enforcedBy: employee_insert/update (manage_org_ids()) and
  // soft_delete_employee(), which re-checks ADMIN-or-MANAGER inside the
  // definer context. Naming who works here is configuring the space, exactly
  // like naming its zones.
  'employees.edit':   ['ADMIN', 'MANAGER'],
  // Running analysis writes telemetry, so it is a configuration act, not a
  // read. A VIEWER watching a live feed would be creating data.
  'analysis.run':     ['ADMIN', 'MANAGER'],

  // ── Governing the organisation. ──
  // enforcedBy: admin_org_ids() — membership_insert/update/delete, org_update.
  // Plus memberships_keep_an_admin, which stops the last admin removing
  // themselves regardless of what any of this says.
  'members.invite':   ['ADMIN'],
  'members.manage':   ['ADMIN'],
  'org.settings':     ['ADMIN'],
};

/** Every capability name, so a typo can be caught rather than silently denied. */
export const ALL_CAPABILITIES = Object.freeze(Object.keys(CAPABILITIES));

/**
 * May this role perform this capability?
 *
 * Fails CLOSED in every ambiguous case:
 *   - unknown role (null, undefined, a suspended member's absent role) -> false
 *   - unknown capability (a typo) -> false
 *
 * The second one matters more than it looks. If `can()` returned true for an
 * unrecognised name, a misspelled capability would silently grant everyone
 * everything, and the mistake would look like working code. Denying instead
 * makes a typo visible as a missing button rather than invisible as a hole.
 */
export function can(role, capability) {
  const allowed = CAPABILITIES[capability];
  if (!allowed) return false;
  if (typeof role !== 'string') return false;
  return allowed.includes(role);
}

/**
 * Every capability a role holds. For passing one object to a client component
 * instead of calling can() a dozen times during render.
 */
export function capabilitiesFor(role) {
  const out = {};
  for (const capability of ALL_CAPABILITIES) {
    out[capability] = can(role, capability);
  }
  return out;
}

/**
 * The message shown when something is refused.
 *
 * Names the role that WOULD be able to do it, because "you do not have
 * permission" leaves the reader with nowhere to go, while "an administrator can
 * do this" tells them exactly who to ask.
 */
export function denialMessage(capability) {
  const allowed = CAPABILITIES[capability];
  if (!allowed) return 'That action is not available.';

  if (allowed.length === 1 && allowed[0] === 'ADMIN') {
    return 'Only an administrator can do that.';
  }
  if (allowed.includes('MANAGER') && !allowed.includes('VIEWER')) {
    return 'Only an administrator or manager can do that.';
  }
  return 'You do not have permission to do that.';
}
